### PR TITLE
Standardize the way fontawesome icons are printed

### PR DESCRIPTION
**Pull request by dregad**
[#27828](https://mantisbt.org/bugs/view.php?id=27828)

### CODE REVIEW

File: core/icon_api.php [hunk 1]
```diff
@@ -34,6 +34,42 @@
...
+ * @param string $p_icon       Fontawesome icon name (fa-xxx)
+ * @param string $p_classes    Optional additional CSS classes
+ * @param string $p_title      Optional pop-up text
+ * @param string $p_inner_html Optional html to insert with the icon.
+ *
+ * @return string
+ */
+function icon_get( $p_icon, $p_classes = '', $p_title = '', $p_inner_html = '' ) {
+	# Add 'fa-' prefix if missing
+	if( substr( $p_icon, 0, 3 ) != 'fa-' ) {
```
Comment (atrol) [inline core/icon_api.php:49]: While this "repairs" bugs in some way, it hides them and adds runtime overhead.
I would remove it.

File: account_prof_menu_page.php [hunk 2]
```diff
@@ -151,12 +151,12 @@
 ?>
 								<td class="center">
 									<?php if( $t_is_global_profile ) { ?>
-									<i class="ace-icon fa fa-check fa-lg"></i>
+									<?php print_icon( 'fa-check', 'ace-icon fa-lg" ); ?>
```
Comment (atrol) [inline account_prof_menu_page.php:154]: Syntax error
`'ace-icon fa-lg"` should be `'ace-icon fa-lg'`

File: account_prof_menu_page.php [hunk 3]
```diff
@@ -151,12 +151,12 @@
...
 									<?php } ?>
 								</td>
 								<td class="center">
 									<?php if( $t_is_default_profile ) { ?>
-									<i class="ace-icon fa fa-check fa-lg"></i>
+									<?php print_icon( 'fa-check', 'ace-icon fa-lg" ); ?>
```
Comment (atrol) [inline account_prof_menu_page.php:159]: Same as above

File: core/layout_api.php [hunk 4]
```diff
@@ -561,13 +561,15 @@ function layout_navbar_button_bar() {
...
-		echo '<i class="fa fa-edit"></i> ' . lang_get( 'report_bug_link' );
+		print_icon( 'fa-edit');
+		echo lang_get( 'report_bug_link' );
 		echo '</a>';
 	}
 
 	if( $t_show_invite_user_button ) {
 		echo '<a class="btn btn-primary btn-sm" href="manage_user_create_page.php">';
-		echo '<i class="fa fa-user-plus"></i> ' . lang_get( 'invite_users' );
+		print_icon( 'fa-fa-user-plus');
```
Comment (atrol) [inline core/layout_api.php:571]: Should be
`print_icon( 'fa-user-plus');`

File: core/layout_api.php [hunk 5]
```diff
@@ -561,13 +561,15 @@ function layout_navbar_button_bar() {
 	if( $t_show_report_bug_button )  {
 		$t_bug_url = string_get_bug_report_url();
 		echo '<a class="btn btn-primary btn-sm" href="' . $t_bug_url . '">';
-		echo '<i class="fa fa-edit"></i> ' . lang_get( 'report_bug_link' );
+		print_icon( 'fa-edit');
+		echo lang_get( 'report_bug_link' );
```
Comment (atrol) [inline core/layout_api.php:565]: Should be `echo ' ' . lang_get( 'report_bug_link' );` to have the same look like before
Now
grafik
Before
grafik

File: roadmap_page.php [hunk 6]
```diff
@@ -96,20 +96,22 @@ function print_version_header( array $p_version_row ) {
...
 	echo '</div>';
 	echo '</div>';
 
 	echo '<div class="widget-body">';
 	echo '<div class="widget-toolbox padding-8 clearfix">';
 	if( $t_scheduled_release_date ) {
-		echo '<div class="pull-left"><i class="fa fa-calendar-o fa-lg"> </i> ' . $t_scheduled_release_date . '</div>';
+		echo '<div class="pull-left">';
+		print_icon( 'fa-calendar-o', 'fa-lg' );
+		echo $t_scheduled_release_date . '</div>';
```
Comment (atrol) [inline roadmap_page.php:114]: Should be `echo ' ' . $t_scheduled_release_date . '</div>';`
Now
grafik
Before
grafik

File: core/layout_api.php [hunk 7]
```diff
@@ -1221,13 +1230,19 @@ function layout_footer() {
...
 			$t_unique_queries_executed = sprintf( lang_get( 'unique_queries_executed' ), $t_unique_queries_count );
-			echo '<small><i class="fa fa-database"></i> ' . $t_unique_queries_executed . '</small>&#160;&#160;&#160;&#160;' . "\n";
+			echo '<small>';
+			print_icon( 'fa-database' );
+			echo ' ' . $t_unique_queries_executed . '</small>&#160;&#160;&#160;&#160;' . "\n";
 		}
 		$t_total_query_time = sprintf( lang_get( 'total_query_execution_time' ), $t_total_query_execution_time );
-		echo '<small><i class="fa fa-clock-o"></i> ' . $t_total_query_time . '</small>&#160;&#160;&#160;&#160;' . "\n";
+		echo '<small>';
+		print_icon( 'fa-fa-clock-o' );
```
Comment (atrol) [inline core/layout_api.php:1244]: Should be `print_icon( 'fa-clock-o' );`

File: account_sponsor_page.php [hunk 8]
```diff
@@ -200,7 +200,7 @@
 		# summary
 		echo '<td>' . string_display_line( $t_bug->summary );
 		if( VS_PRIVATE == $t_bug->view_state ) {
-			printf( ' <i class="fa fa-lock" title="%s"></i>', lang_get( 'private' ) );
+			print_icon( 'fa-lock', '', lang_get( 'private' ) );
```
Comment (atrol) [inline account_sponsor_page.php:203]: Mssing blank after summary
Now
grafik
Before
grafik

File: core/icon_api.php (hunk 1 above)
Comment (dregad) [reply core/icon_api.php:49]: I don't consider this as "repairing a bug".

FontAwesome icons are defined with a name, which is what the API expects. The `fa-` prefix is just a CSS convention, and my idea was to provide an API to deal with _icons_, not CSS classes. So this piece of code is here to provide backwards compatibility.

That said, I hear the argument about overhead, although that is quite negligible IMO. Let's hear what other devs think.

File: account_prof_menu_page.php (hunk 2 above)
Comment (dregad) [reply account_prof_menu_page.php:154]: oops

File: core/layout_api.php (hunk 4 above)
Comment (dregad) [reply core/layout_api.php:571]: OK

File: core/layout_api.php (hunk 5 above)
Comment (dregad) [reply core/layout_api.php:565]: Good catch

File: roadmap_page.php (hunk 6 above)
Comment (dregad) [reply roadmap_page.php:114]: OK

File: core/layout_api.php (hunk 7 above)
Comment (dregad) [reply core/layout_api.php:1244]: That's what happens when using sed for mass search&replace and insufficient double-checking of the changes...

File: account_sponsor_page.php (hunk 8 above)
Comment (dregad) [reply account_sponsor_page.php:203]: OK

**Comment by dregad** [comment]
Thanks @atrol for the thorough review.

File: core/icon_api.php (hunk 1 above)
Comment (dregad) [reply core/icon_api.php:49]: @vboctor your feedback on this would be appreciated.

**Comment by dregad** [comment]
Without further feedback on https://github.com/mantisbt/mantisbt/pull/1725#pullrequestreview-560628602 I will merge the PR as-is; 
I think the overhead of the extra if+substr is negligible, but I'm still open to discussion about this.

**Pull request merged**